controller/blockchain/blockchain.go

```
@@ -1,29 +1,310 @@
+	"github.com/cosmos/go-bip39"
+	"github.com/decred/base58"
 
 	"github.com/SyntropyNet/syntropy-agent/controller"
+	gsrpc "github.com/centrifuge/go-substrate-rpc-client/v3"
+	"github.com/centrifuge/go-substrate-rpc-client/v3/signature"
+	"github.com/centrifuge/go-substrate-rpc-client/v3/types"
+
+	ipfsApi "github.com/ipfs/go-ipfs-api"
+)
+
+const (
+	pkgName        = "Blockchain Controller. "
+	ipfsUrl        = "https://ipfs.io/ipfs/"
+	mnemonicPath   = config.AgentConfigDir + "/mnemonic"
+	addressPath    = config.AgentConfigDir + "/address"
+	reconnectDelay = 10000 // 10 seconds (in milliseconds)
+	waitForMsg     = time.Second
 )
 
+const (
+	// State machine constants
+	stopped = iota
+	connecting
+	running
+)
+
+var ErrNotRunning = errors.New("substrate api is not running")
+
 // Blockchain controller. To be implemented in future
-type BlockchainControler struct {
+type BlockchainController struct {
+	sync.Mutex
+	state.StateMachine
+	substrateApi *gsrpc.SubstrateAPI
+	keyringPair  signature.KeyringPair
+	ipfsShell    *ipfsApi.Shell
+	genesisHash  types.Hash
+	metadata     *types.Metadata
+	comodityKey  types.StorageKey
+	systemKey    types.StorageKey
+
+	url           string
+	lastCommodity []byte
+}
+
+type BlockchainMsg struct {
+	Url string `json:"url"`
+	Cid string `json:"cid"`
 }
 
-var err = errors.New("blockchain controller not yet implemented")
+type Commodity struct {
+	ID      types.Hash
+	Payload []byte
+}
+
+type CommodityInfo struct {
+	Info []byte
+}
+
+func getIpfsPayload(url string) ([]byte, error) {
+	resp, err := http.Get(url)
+	if err != nil {
+		return nil, err
+	}
+
+	defer resp.Body.Close()
+	body, err := io.ReadAll(resp.Body)
+	if err != nil {
+		return nil, err
+	}
+	return body, nil
+}
+
+func getMnemonic() (string, error) {
+	content, err := os.ReadFile(mnemonicPath)
+	if err == nil {
+		return string(content), nil
+	}
+
+	// Mnemonic cache does not exist - create new
+	entropy, _ := bip39.NewEntropy(256)
```

I am wondering why do we swallow error here